### ISSUE TITLE
Variables express different semantics than class composition

### ISSUE THREAD

**Issue by billyjanitsch** [css-modules/css-modules#36]
There's been a lot of discussion surrounding the exclusion of variables in `css-modules`, and there are some good arguments suggesting that composable classes render them obsolete. However, I'd like to elaborate on a case for variables made [here](https://github.com/css-modules/css-modules/issues/22#issuecomment-121477813).

Suppose that I have a primary brand color. Sometimes it's used as a text color, or a background, or a border color, or in a gradient. The only way to do this with (pure) CSS Modules is to make multiple separate classes which each reference that color (with [clever naming](https://github.com/css-modules/css-modules/issues/22#issuecomment-121478537) as suggested by @geelen), and are then composed as needed:

``` css
.primary {
  color: blue;
}
.on-primary {
  background-color: blue;
}
.surrounded-by-primary {
  border-color: blue;
}
// etc.

.myClass {
  composes: on-primary
}
```

This is verbose (both in declaration and in `composes` usage); if the color changes, I have to change all of the references, which leaves a lot of room for error; if I use the color in a new property, I have to introduce an additional reference. It's not unreasonable for a project to have several such colors, which each require this set of classes.

More importantly, it seems like the semantics here are wrong. Composition is not the semantic idea I'm trying to declaratively express. What I actually _am_ expressing is the use of a constant value in different contexts:

``` css
.myClass {
  border-color: $color.primary;
}
```

Even if you find the "X", "on-X", etc. renaming more readable, you still have to declare it for every X, missing the point that X is a variable.

In short, class composition provides a way to reuse property-value pairings, but not a way to reuse values across properties, which I think is fundamentally different (and useful). Thoughts?

(I realize that I can use a preprocessor for variables/mixins, and I respect the goal of keeping CSS Modules bloat free, but it would be really nice if it provided a complete minimal solution where using a preprocessor doesn't feel necessary. This seems like a big enough hole that I can't imagine using CSS Modules alone on any reasonably-sized project.)

P.S. Thanks a ton for your work on this spec!

**Comment by joshwnj** [comment]
@billyjanitsch thanks for your thoughts. I understand you'd like to see this in the core but in the meantime here's an example of using css variables by adding a [postcss plugin](https://github.com/MadLittleMods/postcss-css-variables): 

- https://github.com/x-team/starting-css-modules/tree/master/02-plugins
- https://github.com/x-team/starting-css-modules/blob/master/02-plugins/src/styles/media-item.css#L23

Let me know if that addresses some of the functionality you're looking for.

**Comment by NekR** [comment]
Btw, CSS Modules also is "pre-processing" and if you are not using sass/less/etc. already, you most likely will need to use ``autoprefixer`` (hope you are using it already) and maybe other things which automates such routine things.

But I see your point, here https://github.com/css-modules/css-modules/issues/33#issuecomment-134103295 is my complains about something similar. It actually makes sense to use external processing for variables, but in meantime it feels like we are leaving our comfort zone of CSS Modules usage (because everybody already uses variables and CSS Modules is something better, but we still have to use variables :disappointed: )

**Comment by joshwnj** [comment]
@NekR it might be good at this point in the discussion for us to start looking at real-world examples. Would you like to make a github repo with an example that demonstrates some of the shortcomings you're seeing?  That will be a much more constructive forum where we can try some different approaches and consider the tradeoffs involved.
